### PR TITLE
Scala 2.12 upgrade preview with mysql tests fixed

### PR DESCRIPTION
**Pull request by pankajroark**
Builds on top of #342 fixing the mysql tests that the scalacheck upgrade breaks.

### CODE REVIEW

**Comment by pankajroark** [comment]
All tests have passed afaict, looks like travis is having issues.

File: .travis.yml [hunk 1]
```diff
@@ -2,18 +2,24 @@ language: scala
 sudo: false
 matrix:
   include:
-    - scala: 2.10.6
-      script: umask 0022 && ./sbt ++$TRAVIS_SCALA_VERSION clean test mimaReportBinaryIssues
-
     - scala: 2.11.8
+      jdk: oraclejdk8
```
Comment (ttim) [inline .travis.yml]: Should we do this for 2.11?

File: build.sbt [hunk 2]
```diff
@@ -65,9 +65,9 @@ val ignoredABIProblems = {
 val sharedSettings = extraSettings ++ ciSettings ++ Seq(
   organization := "com.twitter",
   scalaVersion := "2.11.7",
-  crossScalaVersions := Seq("2.10.6", "2.11.7"),
-  javacOptions ++= Seq("-source", "1.6", "-target", "1.6"),
-  javacOptions in doc := Seq("-source", "1.6"),
+  crossScalaVersions := Seq("2.11.7", "2.12.1"),
```
Comment (ttim) [inline build.sbt]: If we add 2.12 here then should we also enable travis for 2.12?

File: storehaus-algebra/src/test/scala/com/twitter/storehaus/algebra/query/CalendarTimeQueryRangeProperties.scala [hunk 3]
```diff
@@ -19,7 +19,8 @@ package com.twitter.storehaus.algebra.query
 import org.scalacheck.{ Gen, Arbitrary }
 import org.scalacheck.Prop._
 import org.scalacheck.Properties
-import com.twitter.scalding._
+import com.twitter.scalding.{RichDate, DateRange, Duration, AbsoluteDuration, Years,
```
Comment (ttim) [inline storehaus-algebra/src/test/scala/com/twitter/storehaus/algebra/query/CalendarTimeQueryRangeProperties.scala]: Do we need it in this review?

**Review by ttim — APPROVED** [review]
Looks good overall, the only questionable thing for me - should we enable 2.12 as part of this change?

File: .travis.yml [hunk 4]
```diff
@@ -2,18 +2,24 @@ language: scala
...
       script: umask 0022 && ./sbt ++$TRAVIS_SCALA_VERSION clean coverage test coverageReport mimaReportBinaryIssues
       after_success:
         - bash <(curl -s https://codecov.io/bash)
+#    - scala: 2.12.1
```
Comment (johnynek) [inline .travis.yml]: what's the story here?

File: build.sbt (hunk 2 above)
Comment (johnynek) [reply build.sbt]: can we use `2.11.11` and `2.12.2`?

File: storehaus-algebra/src/test/scala/com/twitter/storehaus/algebra/query/CalendarTimeQueryRangeProperties.scala (hunk 3 above)
Comment (johnynek) [reply storehaus-algebra/src/test/scala/com/twitter/storehaus/algebra/query/CalendarTimeQueryRangeProperties.scala]: +1 to minimize. I prefer to keep the diffs as minimal as possible (and personally never mix style and logic changes if possible).

File: storehaus-memcache/src/main/scala/com/twitter/storehaus/memcache/MemcacheStore.scala [hunk 5]
```diff
@@ -70,7 +70,6 @@ object MemcacheStore {
       .tcpConnectTimeout(timeout)
       .requestTimeout(timeout)
       .connectTimeout(timeout)
-      .readerIdleTimeout(timeout)
```
Comment (johnynek) [inline storehaus-memcache/src/main/scala/com/twitter/storehaus/memcache/MemcacheStore.scala:73]: this seems like a maybe logical change. Are we sure there is nothing to replace it?

File: storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisSetStore.scala [hunk 6]
```diff
@@ -38,34 +38,34 @@ object RedisSetStore {
...
       case s => Some(s)
     }
 
-  override def put(kv: (ChannelBuffer, Option[Set[ChannelBuffer]])): Future[Unit] =
+  override def put(kv: (Buf, Option[Set[Buf]])): Future[Unit] =
     kv match {
       case (k, Some(v)) =>
-        client.del(Seq(k)) // put, not merge, semantics
-        ttl.map(exp => client.expire(k, exp.inSeconds))
+        client.dels(Seq(k)) // put, not merge, semantics
```
Comment (johnynek) [inline storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisSetStore.scala]: we don't need to wrap with a `Seq`. Could we just call `client.del(k)` maybe? Also, this seems like a a race to me. the del could happen after the set. We should use flatMap to sequence:
```scala
for {
  _ <- client.del(k)
  _ <- ttl.fold(Future.Unit) { exp => client.expire(k, exp.inSeconds) }
  _ <- set(k, v.toList)
} yield ()
```

File: storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisSetStore.scala [hunk 7]
```diff
@@ -38,34 +38,34 @@ object RedisSetStore {
...
+        ttl.map(exp => client.expire(k, exp.inSeconds.toLong))
         set(k, v.toList)
       case (k, None) =>
-        client.del(Seq(k)).unit
+        client.dels(Seq(k)).unit
```
Comment (johnynek) [inline storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisSetStore.scala]: is there no`client.del(k)` we can call?

File: storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisSortedSetStore.scala [hunk 8]
```diff
@@ -32,35 +32,35 @@ object RedisSortedSetStore {
...
+  extends MergeableStore[Buf, Seq[(Buf, Double)]] {
+  def semigroup: Semigroup[Seq[(Buf, Double)]] =
+    implicitly[Semigroup[Seq[(Buf, Double)]]]
 
   /** Returns the whole set as a tuple of seq of (member, score).
    *  An empty set is represented as None. */
-  override def get(k: ChannelBuffer): Future[Option[Seq[(ChannelBuffer, Double)]]] =
-    client.zRange(k, 0, -1, true).map(
+  override def get(k: Buf): Future[Option[Seq[(Buf, Double)]]] =
+    client.zRange(k, 0.toLong, -1.toLong, true).map(
```
Comment (johnynek) [inline storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisSortedSetStore.scala]: `0L` and `-1L` is the better way to do this.

File: storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisSortedSetStore.scala [hunk 9]
```diff
@@ -32,35 +32,35 @@ object RedisSortedSetStore {
...
     )
 
   /** Replaces or deletes the whole set. Setting the set effectivly results
    *  in a delete of the previous sets key and multiple calls to zAdd for each member. */
-  override def put(kv: (ChannelBuffer, Option[Seq[(ChannelBuffer, Double)]])): Future[Unit] =
+  override def put(kv: (Buf, Option[Seq[(Buf, Double)]])): Future[Unit] =
     kv match {
       case (set, Some(scorings)) =>
-        client.del(Seq(set)).flatMap { _ =>
+        client.dels(Seq(set)).flatMap { _ =>
```
Comment (johnynek) [inline storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisSortedSetStore.scala:51]: is there no `client.del(set)`?

File: storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisSortedSetStore.scala [hunk 10]
```diff
@@ -32,35 +32,35 @@ object RedisSortedSetStore {
...
           Future.collect(members.multiPut(scorings.map {
             case (member, score) => ((set, member), Some(score))
           }.toMap).values.toSeq).unit
         }
       case (set, None) =>
-        client.del(Seq(set)).unit
+        client.dels(Seq(set)).unit
```
Comment (johnynek) [inline storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisSortedSetStore.scala:57]: same comment about `client.del(set)`.

File: storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisStore.scala [hunk 11]
```diff
@@ -34,24 +40,33 @@ object RedisStore {
 
   def apply(client: Client, ttl: Option[Duration] = Default.TTL): RedisStore =
     new RedisStore(client, ttl)
+
+  def toChannelBufferStore(store: Store[Buf, Buf]) : Store[ChannelBuffer, ChannelBuffer] = {
+    store.convert(ChannelBuffer2BufInjection.apply)(ChannelBuffer2BufInjection)
+  }
+}
+
+object ChannelBuffer2BufInjection extends Injection[ChannelBuffer, Buf] {
```
Comment (johnynek) [inline storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisStore.scala]: this should be a bijection, not an Injection (there is an implicit from Bijection to Injection, by the way so you can pass this as the injection above).

File: storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisSortedSetStore.scala (hunk 9 above)
Comment (pankajroark) [reply storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisSortedSetStore.scala:51]: There is but it's marked deprecated.

File: storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisSetStore.scala (hunk 7 above)
Comment (pankajroark) [reply storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisSetStore.scala]: del is deprecated:
  @deprecated("remove netty3 types from public API", "2016-03-24")
  def del(keys: Seq[ChannelBuffer]): Future[JLong] =

File: storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisSetStore.scala (hunk 6 above)
Comment (pankajroark) [reply storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisSetStore.scala]: del is deprecated unfortunately. Taken care of the race.

File: storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisSortedSetStore.scala (hunk 10 above)
Comment (pankajroark) [reply storehaus-redis/src/main/scala/com/twitter/storehaus/redis/RedisSortedSetStore.scala:57]: del is deprecated.

File: .travis.yml (hunk 1 above)
Comment (pankajroark) [reply .travis.yml]: No, I guess in the original PR Ben may have added it for consistency. Let me remove it.

File: .travis.yml (hunk 4 above)
Comment (pankajroark) [reply .travis.yml]: Ahh... this got carried over from Ben's original PR. Let me uncomment this.

File: storehaus-algebra/src/test/scala/com/twitter/storehaus/algebra/query/CalendarTimeQueryRangeProperties.scala (hunk 3 above)
Comment (pankajroark) [reply storehaus-algebra/src/test/scala/com/twitter/storehaus/algebra/query/CalendarTimeQueryRangeProperties.scala]: Removed.

File: storehaus-memcache/src/main/scala/com/twitter/storehaus/memcache/MemcacheStore.scala (hunk 5 above)
Comment (pankajroark) [reply storehaus-memcache/src/main/scala/com/twitter/storehaus/memcache/MemcacheStore.scala:73]: Applied the alternative way of doing this.

**Comment by oscar-stripe** [comment]
+1

After we publish a new version, we should bump the minor version, and remove all the exclusions for mima.

Thanks for doing this.

**Pull request merged**